tests/test_session_0.py

```diff
@@ -0,0 +1,185 @@
+        # collection before with no luck. This block fails even with
+        # "import calendar" before it. Needed to restore the original modules
+        # with the 'copy_modules' object. (Moved to "test_session_byref_*.py".)
+        #for obj in ('Calendar', 'isleap'):
+        #    assert main_dict[obj] is sys.modules['calendar'].__dict__[obj]
+        #assert main_dict['day_name'].__module__ == 'calendar'
+        #if byref:
+        #    assert main_dict['day_name'] is sys.modules['calendar'].__dict__['day_name']
+
+        for obj in ('x', 'empty', 'names'):
+            assert main_dict[obj] == copy_dict[obj]
+
+        globs = '__globals__' if dill._dill.PY3 else 'func_globals'
+        for obj in ['squared', 'cubed']:
+            assert getattr(main_dict[obj], globs) is main_dict
+            assert main_dict[obj](3) == copy_dict[obj](3)
+
+        assert main.Person.__module__ == main.__name__
+        assert isinstance(main.person, main.Person)
+        assert main.person.age == copy_dict['person'].age
+
+        assert issubclass(main.CalendarSubclass, main.Calendar)
+        assert isinstance(main.cal, main.CalendarSubclass)
+        assert main.cal.weekdays() == copy_dict['cal'].weekdays()
+
+        assert main.selfref is main
+
+    except AssertionError:
+        import traceback
+        error_line = traceback.format_exc().splitlines()[-2].replace('[obj]', '['+repr(obj)+']')
+        print("Error while testing (byref=%s):" % byref, error_line, sep="\n", file=sys.stderr)
+        raise
+
+
+if __name__ == '__main__':
+
+    # Test dump_session() and load_session().
+    for byref in (False, True):
+        if byref:
+            # Test unpickleable imported object in main.
+            try:
+                from ctypes import pythonapi
+            except ImportError:
+                pass
+
+        #print(sorted(set(sys.modules.keys()) - original_modules))
+        dill._test_file = dill._dill.StringIO()
+        try:
+            # For the following test files.
+            dill.dump_session('session-byref-%s.pkl' % byref, byref=byref)
+
+            dill.dump_session(dill._test_file, byref=byref)
+            dump = dill._test_file.getvalue()
+            dill._test_file.close()
+
+            import __main__
+            copy_dict = __main__.__dict__.copy()
+            copy_modules = sys.modules.copy()
+            del copy_dict['dump']
+            del copy_dict['__main__']
+            for name in copy_dict.keys():
+                if name not in original_objects:
+                    del __main__.__dict__[name]
+            for module in list(sys.modules.keys()):
+                if module not in original_modules:
+                    del sys.modules[module]
+
+            dill._test_file = dill._dill.StringIO(dump)
+            dill.load_session(dill._test_file)
+            #print(sorted(set(sys.modules.keys()) - original_modules))
+        finally:
+            dill._test_file.close()
```

Comment:
should remove `*pkl` session files here as cleanup, if finishes correctly or errors out

Reply:
solved